src/levanter/models/whisper.py

```diff
@@ -0,0 +1,562 @@
+        )
+
+    @classmethod
+    def from_hf_config(cls, hf_config: HfConfig):
+        return cls(
+            vocab_size=hf_config.vocab_size,
+            num_mel_bins=hf_config.num_mel_bins,
+            encoder_layers=hf_config.encoder_layers,
+            encoder_attention_heads=hf_config.encoder_attention_heads,
+            decoder_layers=hf_config.decoder_layers,
+            decoder_attention_heads=hf_config.decoder_attention_heads,
+            decoder_ffn_dim=hf_config.decoder_ffn_dim,
+            encoder_ffn_dim=hf_config.encoder_ffn_dim,
+            activation_function=hf_config.activation_function,
+            max_source_positions=hf_config.max_source_positions,
+            d_model=hf_config.d_model,
+        )
+
+
+class WhisperMlp(eqx.Module, StateDictSerializationMixin):
+    fc1: hnn.Linear  # projection from Embed to Intermediate (typically 4x Embed)
+    fc2: hnn.Linear  # projection from Intermediate to Embed
+    act: Callable = eqx.static_field()
+
+    @staticmethod
+    def init(Embed: Axis, Mlp: Axis, activation_fn, *, key, use_bias: bool = True) -> "WhisperMlp":
+        k_fc, k_proj = haliax.jax_utils.maybe_rng_split(key, 2)
+        fc1 = hnn.Linear.init(Out=Mlp, In=Embed, key=k_fc, use_bias=use_bias)
+        fc2 = hnn.Linear.init(Out=Embed, In=Mlp, key=k_proj, use_bias=use_bias)
+        if isinstance(activation_fn, str):
+            activation_fn = ACT2FN[activation_fn]
+        act = activation_fn  # type: ignore
+
+        return WhisperMlp(fc1, fc2, act)
+
+    @named_call
+    def __call__(self, x: NamedArray, *, key=None):
+        k1, k2 = haliax.jax_utils.maybe_rng_split(key, 2)
+        x = self.fc1(x, key=k1)
+        x = self.act(x)
+        x = self.fc2(x, key=k2)
+        return x
+
+    def from_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None):
+        d = {}
+        d.update(
+            unflatten_linear_layers(apply_prefix(prefix, "fc1"), state_dict, self.fc1, out_dims_first_in_dict=True)
+        )
+        d.update(
+            unflatten_linear_layers(apply_prefix(prefix, "fc2"), state_dict, self.fc2, out_dims_first_in_dict=True)
+        )
+
+        return super().from_state_dict(d, prefix)
+
+    def update_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None) -> StateDict:
+        my_dict: StateDict = {}
+        super().update_state_dict(my_dict, prefix=prefix)
+
+        my_dict.update(flatten_linear_layers(apply_prefix(prefix, "fc1"), self.fc1, out_dims_first_in_dict=True))
+        my_dict.update(flatten_linear_layers(apply_prefix(prefix, "fc2"), self.fc2, out_dims_first_in_dict=True))
+
+        state_dict.update(my_dict)
+        return state_dict
+
+
+class WhisperAttention(StateDictSerializationMixin, eqx.Module):
+    config: WhisperConfig = eqx.static_field()
+
+    q_proj: hnn.Linear  # input projection from [embed] -> [q, heads, head_dim]
+    # Unlike other attention blocks in Levanter, splits query to support cross-attn
```

Comment:
it's really only gpt2 that does it this way still i think fwiw

Reply:
I'm actually somewhat curious if grouping all the linears has perf advantages with Jax. Grouping the linear layers has perf advantages in Pytorch, but does the compiler of jax recognize these can be run in parallel and do so?

Reply:
I guess the cost of these linear layers is relatively minimal v.s. the MLP for most models so it might not matter much.

Reply:
Removed the comment regardless though!

Reply:
I would be shocked if it mattered on TPU especially, but JAX is pretty good at fusing and overlapping things. I wonder if torch.compile also can make most of the difference go away?